draft-ietf-quic-tls.md

```diff
@@ -724,6 +703,27 @@ This results in abandoning loss recovery state for the Initial encryption level
 and ignoring any outstanding Initial packets.
 
 
+### Discarding Handshake Keys
+
+An endpoint MUST NOT discard its handshake keys until the TLS handshake is
+confirmed ({{handshake-confirmed}}).  An endpoint SHOULD discard its handshake
+keys as soon as it has confirmed the handshake.  Endpoints SHOULD repeatedly
+send ACK-eliciting frames encrypted with 1-RTT keys as soon as they have
+installed those send keys and until they receive an acknowledgment for one of
+them.
+
+
+### Discarding 0-RTT Keys
+
+Clients SHOULD discard 0-RTT keys as soon as they install 1-RTT keys, since
+they have no use after that moment.
+
+Servers MUST discard 0-RTT keys as soon as they no longer expect to receive
+0-RTT packets.  This can been established by receiving a 1-RTT packet when all
+previous packet numbers have been received, or by receiving a 1-RTT packet and
+then waiting up to three times the Probe Timeout (PTO, see {{QUIC-RECOVERY}}).
```

Comment:
I would rewrite this.

"Once a 1-RTT packet has been received, then no 0-RTT packets with larger packet numbers will be received (assuming a compliant client). Upon receiving the first 1-RTT packet, servers MUST discard the 0-RTT no later than three times the Probe Timeout. If a server tracks all the 0-RTT packet numbers, it can safely discard the 0-RTT keys when all packets up to the first 1-RTT packets have been received. Note that even if the 0-RTT keys are discarded prematurely, the packets will eventually retransmitted".

Reply:
I don't think I prefer your wording. I think this is editorial and will let the editors wordsmith to this level of detail

Reply:
This isn't quite editorial.  I have a few concerns with ekr's proposed text, but it has some useful pieces.  With a rewrite...

1. Context.  The statement: "A compliant client will not send 0-RTT packets after it starts sending 1-RTT packets." is relevant.

2. Reordering happens: "After receiving a 1-RTT packet, a server might need to retain 0-RTT keys to allow for reordered packets.  Retaining keys for up to three times the Probe Timeout (PTO, ...) ensures that packets are not dropped as a result of prematurely discarding keys."

3. And this is only a maybe for me: "A server can discard keys earlier once it receives all 0-RTT packets."

Reply:
I think MT's proposal is basically OK. I would prefer to include something like the last sentence to explain why 3PTO is just a compromise and not  a perfect guarantee of safety, but also that it doesn't create a deadlock or lost data if you are too aggressive.

Reply:
3PTO seems way too long. The server only needs to keep the keys around to match out of order delivery delays. The PTO itself is defined in the recovery spec as:
```
PTO = smoothed_rtt + max(4*rttvar, kGranularity) + max_ack_delay
```
In that formula, the middle component accounts for out of order delivery. If we believe that, the server only needs to keep the 0-RTT keys around for `max(4*rttvar, kGranularity)`.

Note that dropping packets is not the end of the world -- dropped packets will be retransmitted. If the ACK of the 0-RTT packet is not received by the client 1 PTO after it was sent, the content will be retransmitted in an 1-RTT packet. That retransmitted packet will be received at most 1 RTT after the first 1-RTT packet, which means the server has no reason to keeping the 0-RTT keys around longer than 1-RTT after the first 1-RTT packet is received.

Reply:
@huitema I am not sure if using `max(4*rttvar, kGranularity)` is the correct thing to do, because the impact of reordering going above the threshold is different.

At the moment `max(4*rttvar, kGranularity)` is used to trigger a retransmission. The effect is wasting 1 MTU at most (there would be no loss at all if the sender choose something other to send than the packet that was deemed lost). Dropping key leads to a packet that leads to a real packet loss. That'd have negative impact on CC.

Therefore, I'd prefer having a conservative recommendation that `max(4*rttvar, kGranularlity)`, and reusing 3PTO here seems like a good choice, even though I agree that it might be longer than necessary.

Reply:
I've rephrased the text in such a way that the context is clearer and the server can discard its 0-RTT keys at any point between receiving the first 1-RTT packet, and 3*PTO after that. This allows implementors to decide how long they wish to wait.